### PR TITLE
fix provide isn't reactive with a single array

### PR DESCRIPTION
**Pull request by Kingwl**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
#5223

### CODE REVIEW

**Comment by HerringtonDarkholme** [comment]
Related: https://github.com/vuejs/vue/issues/5067 

Would @znck please also give a look?

**Comment by znck** [comment]
One thing I don't want with provide/inject is child modifying ancestor's data.

If `foo` is injected & reactive then setting `this.foo = null` would set the provided prop to `null`. IMO that could be problematic.

Injected properties were kept non-reactive intentionally. Perhaps @yyx990803 could explain better.

**Comment by Kingwl** [comment]
https://facebook.github.io/react/docs/context.html#updating-context

> React has an API to update context, but it is fundamentally broken and you should not use it.

should we warn when the context is changed？

**Comment by znck** [comment]
Currently, reactive objects can be provided. 

https://jsfiddle.net/m0fyr6z1/3/ - In this example, you see, child re-renders if array is changed in parent component.

IMO, rather than setting provided injections (current implementation) or defining reactive injections (this PR), it would be better to add enumerable, non-configurable & read-only injections.

**Comment by yyx990803** [comment]
`defineReactive` accepts a fourth argument (customSetter) which allows overriding the default set behavior. We can just use that to do the warning when a child component attempts to mutate an injection binding. This is what we are doing for props too: https://github.com/vuejs/vue/blob/dev/src/core/instance/state.js#L71-L91

File: src/core/instance/inject.js [hunk 1]
```diff
@@ -29,7 +31,14 @@ export function initInjections (vm: Component) {
       let source = vm
       while (source) {
         if (source._provided && provideKey in source._provided) {
-          vm[key] = source._provided[provideKey]
+          defineReactive(vm, key, source._provided[provideKey], () => {
+            warn(
```
Comment (yyx990803) [inline src/core/instance/inject.js]: make sure to prefix warnings with `process.env.NODE_ENV` check :)

Comment (Kingwl) [inline src/core/instance/inject.js]: done

**Comment by rayrutjes** [comment]
Overriding the setter makes this PR introduce a BC.

**Comment by yyx990803** [comment]
It doesn't because it was not reactive anyway.

**Comment by rayrutjes** [comment]
In the project I'm working on, I did the following:

```js
export default {
  inject: ['_searchStore'],
  props: {
    searchStore: {
      type: Object,
      default () {
        return this._searchStore
      }
    }
  }
}
```

Which will now break if I'm not mistaken.

**Pull request merged**